### PR TITLE
Update linter config

### LINKED ISSUE #132: [CI] Update linter settings
**Location**: `.golangci.yml`  

**Problem**: The current linter settings do not reflect our programming style anymore.  
The config is also slightly outdated.

Observed problems:  
- [x] Undocumented exported  types/funcs (doen with `revive`)
- [ ] No `Get…` prefix warning (to be discussed, would use `forbidgo`)
- [x] Check copyright header (done with `goheader`)
- [x] https://github.com/perun-network/poly-go/pull/11#discussion_r749212486

**Solution**: Take a look https://golangci-lint.run/usage/linters/ and evaulate which linters we need and update `.golangci.yml`.  
In order to merge the changed linter settings, all warnings would need to be fixed.

### PR DESCRIPTION
**Pull request by ggwpez**
- Fixes linter warnings
- Updates the linter config

Closes #132  
Depends on #256

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -9,95 +9,54 @@ run:
...
+    - maligned
+    - scopelint
+    - golint
+    - interfacer
+
+linters-settings:
+  gocritic:
+    disabled-checks:
+      # Accept code in the form of object.Call(object), eg:
+      # assert.True(t, null.Equals(null), "Expected equality of zero address to itself")
```
Comment (matthiasgeihs) [inline .golangci.yml:44]: Why would we want to allow that in general? This is an instance where I think it makes sense to disable the linter for a specific line.

File: .golangci.yml [hunk 2]
```diff
@@ -9,95 +9,54 @@ run:
...
+      - dupArg
+  cyclop:
+    # 15 is used by go-report-card as well.
+    max-complexity: 15
+    skip-tests: true
```
Comment (matthiasgeihs) [inline .golangci.yml:49]: Why skip tests?

File: .golangci.yml [hunk 3]
```diff
@@ -9,95 +9,54 @@ run:
...
+    - gofumpt # Pedantic formatting
+    - revive # go lint drop in
+    - ifshort # we often dont use `if err := …` for readability.
+    - tparallel # We dont always use parallel tests.
+
+    - wrapcheck # We do not rewrap all errors from other packages.
+    - varnamelen # We often use single letter arguments.
+    - exhaustivestruct # We often have uninitialized fields of structs.
+    - errorlint # We have interface errors which an impl needs to return.
+    - nakedret # We use naked returns in long functions.
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply .golangci.yml:24]: Ok i was under the impression that this always enforced named returns on functions, but it only enforces to use the return names when the function has named returns.  
It are 6 occurrences, i will fix them.

[the remark this answers is not in the document]
Comment (ggwpez) [reply .golangci.yml:24]: Somehow this linter only reports it at the end of functions, anyway its fixed now.

File: .golangci.yml [hunk 4]
```diff
@@ -9,95 +9,54 @@ run:
...
+    - lll # We use long lines.
+    - gochecknoglobals # We have a lot of globals.
+    - nlreturn # "new line before return"; not useful.
+    - paralleltest # We like parallel tests.
+    - ireturn # We often return interfaces.
+    - wsl # Formatting disabled for now.
+    - gochecknoinits # We use init functions.
+    - gci # We have our own import order
+    # Unimportant
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply .golangci.yml:33]: Ups, i removed it.

File: .golangci.yml (hunk 1 above)
Comment (ggwpez) [reply .golangci.yml:44]: Ok i updated it.

**Comment by ggwpez** [comment]
I addressed the comments. There are two points for discussion remaining:  
1. Do we want to forbid functions to start with `Get…`?
2. Enabling `revive` (formerly golint) reports:
```pre
exported: func name will be used as test.TestAddress by other packages, and that stutters; consider calling this Address
exported: func name will be used as test.TestAccountWithWalletAndBackend by other packages, and that stutters; consider calling this AccountWithWalletAndBackend
```
I think this is the reason why we normally named the generic tests with `Generic…Test`.  You can see the reported errors [here](https://github.com/hyperledger-labs/go-perun/runs/4225465309?check_suite_focus=true#step:4:73).
@matthiasgeihs

**Comment by matthiasgeihs** [comment]
> 1. Do we want to forbid functions to start with `Get…`?

Typically we shouldn't use "Get...". I had a look and it seems like there are easy work arounds for the locations where we are using it. In many cases we can simply remove it.

> 2. Enabling `revive` (formerly golint) reports:
> 
> ```
> exported: func name will be used as test.TestAddress by other packages, and that stutters; consider calling this Address
> exported: func name will be used as test.TestAccountWithWalletAndBackend by other packages, and that stutters; consider calling this AccountWithWalletAndBackend
> ```
> 
> I think this is the reason why we normally named the generic tests with `Generic…Test`.

I see. However, I don't think "Generic...Test" is a good name (or idiomatic) either. I don't have the perfect solution for this. I see two options:
- Either we surpress the linter warning for the respective cases.
- We go back to the other naming schema.

I have a preference for 1, but as I said, it is also not perfect. Any other suggestions?

**Comment by ggwpez** [comment]
> Typically we shouldn't use "Get...". I had a look and it seems like there are easy work arounds for the locations where we are using it. In many cases we can simply remove it.  

Yes i think so as well.

> I have a preference for 1, but as I said, it is also not perfect. Any other suggestions?

I dont mind, but i also dont have a better naming scheme :laughing: will suppress for now.

**Comment by ggwpez** [comment]
All addressed. Changes are in the last [force push](https://github.com/hyperledger-labs/go-perun/compare/18effa18e75539a24d7ed91751af08e24c302668..d02fd58130ec8e98601eb01e863d3cf195e5d79d).  
@matthiasgeihs

File: wire/net/simple/listener_internal_test.go [hunk 5]
```diff
@@ -80,7 +80,7 @@ func TestListener_Accept(t *testing.T) {
 		defer l.Close()
 
 		test.AssertNotTerminates(t, timeout, func() {
-			l.Accept()
+			_, _ = l.Accept()
```
Comment (matthiasgeihs) [inline wire/net/simple/listener_internal_test.go:83]: Do something with the return values or use `//nolint`?

File: apps/payment/app_internal_test.go [hunk 6]
```diff
@@ -38,9 +38,9 @@ func TestApp_ValidInit(t *testing.T) {
 	app := new(App)
 
 	nildata := &channel.State{Data: nil}
-	assert.Panics(func() { app.ValidInit(nil, nildata) })
+	assert.Panics(func() { _ = app.ValidInit(nil, nildata) })
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply apps/payment/app_internal_test.go:41]: We talked about the use of `_ =` vs `// nolint` on monday with the example of  
`defer func() { _ = closer.Closer() } ()`  
as opposed to
`defer closer.Closer() // nolint: errcheck`  
and came to the conclusion that suppressing was worse.  
Which is why i now always used `_ = ` instead of suppressing the linter.

File: backend/ethereum/channel/adjudicator_test.go [hunk 7]
```diff
@@ -36,6 +36,7 @@ import (
 const defaultTxTimeout = 2 * time.Second
 
 func testSignState(t *testing.T, accounts []*keystore.Account, state *channel.State) channel.Transaction {
+	t.Helper()
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/channel/adjudicator_test.go:39]: The doc of `t.Helper()` states this:  
```pre
Helper marks the calling function as a test helper function. When printing file and line information, that function will be skipped. Helper may be called simultaneously from multiple goroutines.
```
and was done to satisfy the `thelper` linter.

File: backend/ethereum/channel/backend.go [hunk 8]
```diff
@@ -226,7 +226,10 @@ func EncodeState(state *adjudicator.ChannelState) ([]byte, error) {
 func assetToCommonAddresses(addr []channel.Asset) []common.Address {
 	cAddrs := make([]common.Address, len(addr))
 	for i, part := range addr {
-		asset := part.(*Asset)
+		asset, ok := part.(*Asset)
+		if !ok {
+			log.Panic("wrong address type")
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/channel/backend.go:231]: updated

File: backend/ethereum/channel/contractbackend.go [hunk 9]
```diff
@@ -33,8 +33,11 @@ import (
 	pcontext "polycry.pt/poly-go/context"
 )
 
-// How many blocks we query into the past for events.
-const startBlockOffset = 100
+const (
+	// How many blocks we query into the past for events.
+	startBlockOffset            = 100
+	contractBackendHeadBuffSize = 10
+)
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/channel/contractbackend.go:40]: moved it up

File: backend/sim/wallet/address.go [hunk 10]
```diff
@@ -68,33 +70,38 @@ func (a *Address) Bytes() []byte {
...
-	copy(data[64-len(yb):64], yb)
+	copy(data[addrLen/2-len(xb):addrLen/2], xb)
+	copy(data[addrLen-len(yb):addrLen], yb)
 
 	return data
 }
 
 // String converts this address to a human-readable string.
 func (a *Address) String() string {
+	const bytePerNibble = 4
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/sim/wallet/address.go:86]: The sim addresses have shorter encodings for easier debugging when reading the logs.  
Updated it to `length`.

File: channel/persistence/keyvalue/restorer.go [hunk 11]
```diff
@@ -92,7 +92,7 @@ func (pr *PersistRestorer) RestorePeer(addr wire.Address) (persistence.ChannelIt
 		return nil, errors.WithMessage(err, "restoring peer")
 	}
 	itPeer := sortedkv.NewTable(pr.db, prefix.PeerDB+key).NewIterator()
-	defer itPeer.Close() // nolint: errcheck
+	defer itPeer.Close()
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply channel/persistence/keyvalue/restorer.go:95]: I dont know, seems that `golangci-lint` does not report this anymore. Their changelog is unreadable :shrug:

File: channel/transaction.go [hunk 12]
```diff
@@ -61,9 +61,13 @@ func (t *Transaction) Decode(r io.Reader) error {
...
 	}
-	if (stateSet % 2) == 0 {
+	switch stateSet {
+	case 0:
 		t.State = nil
 		return nil
+	case 1:
+	default:
+		return errors.Errorf("unknown stateSet value: %v", stateSet)
 	}
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply channel/transaction.go:71]: I wanted to get rid of the magic number `2`, but then i saw that the code should better reject values different from `0` and `1` since they would decode incorrectly anyway.

File: channel/transaction_test.go [hunk 13]
```diff
@@ -26,7 +26,7 @@ import (
 func TestTransactionSerialization(t *testing.T) {
 	rng := pkgtest.Prng(t)
 	lengths := []int{2, 5, 10, 40}
-	var tests [][]bool
+	var tests [][]bool // nolint: prealloc
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply channel/transaction_test.go:29]: Since i dont know how to calculate the length, would be something like `len(lengths) * 4 + ?`.

File: client/proposal.go [hunk 14]
```diff
@@ -687,15 +692,15 @@ func (c *Client) enableVer1Cache() {
...
+func (c *Client) releaseVer1Cache(ctx context.Context) {
 	c.log.Trace("Releasing version 1 cache")
 
 	c.version1Cache.mu.Lock()
 	defer c.version1Cache.mu.Unlock()
 
 	c.version1Cache.enabled--
 	for _, u := range c.version1Cache.cache {
-		go c.handleChannelUpdate(u.uh, u.p, u.m)
+		go c.handleChannelUpdate(ctx, u.uh, u.p, u.m)
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/proposal.go:703]: This is done for the [contextcheck](https://github.com/sylvia7788/contextcheck) linter.

File: backend/ethereum/wallet/backend.go [hunk 15]
```diff
@@ -31,6 +32,10 @@ type Backend struct{}
 // ref https://github.com/ethereum/go-ethereum/blob/54b271a86dd748f3b0bcebeaf678dc34e0d6177a/crypto/signature_cgo.go#L66
 const SigLen = 65
 
+// SigVSubtract value that is subtracted from the last byte of a signature if
+// the last bytes exceeds it.
+const SigVSubtract = 27
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/wallet/backend.go:37]: It is now used in `backend/eth/chan/test/transactor.go sigFromRSV` function and would need to be public for that.

File: client/channelconn.go [hunk 16]
```diff
@@ -42,11 +42,11 @@ type channelConn struct {
 
 // newChannelConn creates a new channel connection for the given channel ID. It
 // subscribes on the subscriber to all messages regarding this channel.
-func newChannelConn(id channel.ID, peers []wire.Address, idx channel.Index, sub wire.Subscriber, pub wire.Publisher) (_ *channelConn, err error) {
+func newChannelConn(_ context.Context, id channel.ID, peers []wire.Address, idx channel.Index, sub wire.Subscriber, pub wire.Publisher) (_ *channelConn, err error) {
 	// relay to receive all update responses
 	relay := wire.NewRelay()
 	// we cache all responses for the lifetime of the relay
-	relay.Cache(context.Background(), func(*wire.Envelope) bool { return true })
+	relay.Cache(context.Background(), func(*wire.Envelope) bool { return true }) // nolint: contextcheck
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/channelconn.go:49]: The `contextcheck` linter wants to always have context passed down to a function when it uses one.  
In this case the context would be to short. I think `relay.Cache` should not accept a context anyway.

File: client/client_persistence_test.go [hunk 17]
```diff
@@ -27,17 +27,18 @@ func TestPersistencePetraRobert(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), twoPartyTestTimeout)
 	defer cancel()
 
-	runTwoPartyTest(ctx, t, func(rng *rand.Rand) (setups []ctest.RoleSetup, roles [2]ctest.Executer) {
+	runTwoPartyTest(t, ctx, func(rng *rand.Rand) (setups []ctest.RoleSetup, roles [2]ctest.Executer) {
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/client_persistence_test.go:30]: Was done for the `thelper` linter, but now it does not report an error if i change it back. Reverting.

File: client/restore.go [hunk 18]
```diff
@@ -24,21 +24,23 @@ import (
 	"perun.network/go-perun/wire"
 )
 
-type channelFromSourceSig = func(*Client, *persistence.Channel, *Channel, ...wire.Address) (*Channel, error)
+type channelFromSourceSig = func(context.Context, *Client, *persistence.Channel, *Channel, ...wire.Address) (*Channel, error)
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/restore.go:27]: Seems that one of the functions that call `channelFromSourceSig` would otherwise not pass it down.

File: client/restore_internal_test.go [hunk 19]
```diff
@@ -32,7 +34,11 @@ import (
 	pkgtest "polycry.pt/poly-go/test"
 )
 
+// timeout for reconstructChannel calls.
+var reconstructTimeout = 10 * time.Second
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/restore_internal_test.go:38]: From all the linters - no one catches this :facepalm: :hankey:

File: client/sync.go [hunk 20]
```diff
@@ -71,7 +70,6 @@ func (c *Client) handleSyncMsg(peer wire.Address, msg *msgChannelSync) {
 // nolint:unused
 func (c *Client) syncChannel(ctx context.Context, ch *persistence.Channel, p wire.Address) (err error) {
 	recv := wire.NewReceiver()
-	// nolint:errcheck
 	defer recv.Close() // ignore error
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/sync.go:73]: I dont know, `golangci-lint` does not report it anymore.

File: client/sync.go [hunk 21]
```diff
@@ -101,7 +99,10 @@ func (c *Client) syncChannel(ctx context.Context, ch *persistence.Channel, p wir
 	if err != nil {
 		return errors.WithMessage(err, "receiving sync message")
 	}
-	msg := env.Msg.(*msgChannelSync) // safe by the predicate
+	msg, ok := env.Msg.(*msgChannelSync) // safe by the predicate
+	if !ok {
+		log.Panic("internal error: wrong message type")
+	}
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/sync.go:105]: We have a message filter predicate above, so the type should be correct already. I removed the comment.

File: client/test/alice.go [hunk 22]
```diff
@@ -33,8 +33,9 @@ type Alice struct {
 }
 
 // NewAlice creates a new Proposer that executes the Alice protocol.
-func NewAlice(setup RoleSetup, t *testing.T) *Alice {
-	return &Alice{Proposer: *NewProposer(setup, t, 3)}
+func NewAlice(t *testing.T, setup RoleSetup) *Alice {
+	t.Helper()
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/test/alice.go:37]: For the `thelper` linter since it is not a test function itself but just a helper which accepts a `*testing.T`.

File: client/test/backend.go [hunk 23]
```diff
@@ -360,8 +360,8 @@ func (b *MockBackend) removeSubscription(ch channel.ID, sub *mockSubscription) {
 
 	// Find subscription index.
 	i, ok := func() (int, bool) {
-		for i, sub_ := range b.eventSubs[ch] {
-			if sub == sub_ {
+		for i, s := range b.eventSubs[ch] {
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/test/backend.go:363]: The `revive` linter does not like underscores in names.

File: client/update.go [hunk 24]
```diff
@@ -209,7 +207,10 @@ func (c *Channel) updateGeneric(
 		return newPeerRejectedError("channel update", rej.Reason)
 	}
 
-	acc := res.(*msgChannelUpdateAcc) // safe by predicate of the updateResRecv
+	acc, ok := res.(*msgChannelUpdateAcc) // safe by predicate of the updateResRecv
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/update.go:210]: The predicate of `NewUpdateResRecv` already filters the message type, so there should only `msgChannelUpdateAcc` arrive here.

File: client/virtual_channel.go [hunk 25]
```diff
@@ -74,37 +77,38 @@ func (c *Channel) proposeVirtualChannelFunding(ctx context.Context, virtual *Cha
 	return err
 }
 
-const responseTimeout = 10 * time.Second          // How long we wait until the proposal response must be transmitted.
-const virtualFundingTimeout = 10 * time.Second    // How long we wait for a matching funding proposal.
-const virtualSettlementTimeout = 10 * time.Second // How long we wait for a matching settlement proposal.
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/virtual_channel.go:79]: Removed; it was unused.

File: log/log.go [hunk 26]
```diff
@@ -39,8 +39,8 @@ func Set(l Logger) {
 	logger = l
 }
 
-// Get returns the currently set framework logger.
-func Get() Logger {
+// Log returns the currently set framework logger.
+func Log() Logger {
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply log/log.go:43]: Ok i dont mind.

File: log/log.go [hunk 27]
```diff
@@ -54,11 +54,9 @@ type StdLogger interface {
 	Printf(format string, args ...interface{})
 	Print(...interface{})
 	Println(...interface{})
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply log/log.go:56]: The `StdLogger` has the same interface as the `stdlog.Log` struct. But that is quite the stretch if someone only looks at our types :laughing:

File: wallet/sig.go [hunk 28]
```diff
@@ -82,18 +85,19 @@ func EncodeSparseSigs(w io.Writer, sigs []Sig) error {
...
 		return errors.WithMessage(err, "decoding mask")
 	}
 
+	const two = 2
 	// Decoding mask's signatures
 	for maskIdx, sigIdx := 0, 0; maskIdx < len(mask); maskIdx++ {
-		for bitIdx := 0; bitIdx < 8 && sigIdx < len(*sigs); bitIdx, sigIdx = bitIdx+1, sigIdx+1 {
-			if ((mask[maskIdx] >> bitIdx) % 2) == 0 {
+		for bitIdx := 0; bitIdx < bitPerByte && sigIdx < len(*sigs); bitIdx, sigIdx = bitIdx+1, sigIdx+1 {
+			if ((mask[maskIdx] >> bitIdx) % two) == 0 {
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply wallet/sig.go:100]: Ok renamed to `binaryModulo`.

File: wallet/sig.go [hunk 29]
```diff
@@ -59,10 +61,11 @@ func EncodeSparseSigs(w io.Writer, sigs []Sig) error {
 	n := len(sigs)
 
 	// Encode mask
-	mask := make([]uint8, int(math.Ceil(float64(n)/8.0)))
+	const one = 1
+	mask := make([]uint8, int(math.Ceil(float64(n)/float64(bitPerByte))))
 	for i, sig := range sigs {
 		if sig != nil {
-			mask[i/8] |= 0x01 << (i % 8)
+			mask[i/bitPerByte] |= one << (i % bitPerByte)
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply wallet/sig.go:68]: Ok i applied a `nolint`.

File: wallet/test/address.go [hunk 30]
```diff
@@ -24,7 +24,7 @@ import (
 
 // TestAddress runs a test suite designed to test the general functionality of
 // an address implementation.
-func TestAddress(t *testing.T, s *Setup) {
+func TestAddress(t *testing.T, s *Setup) { // nolint: revive
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply wallet/test/address.go:27]: I added
```pre
//nolint:revive
// func name will be used as test.TestAddress by other packages,
// and that stutters; consider calling this Address (revive).
```

File: wallet/test/wallet.go [hunk 31]
```diff
@@ -43,7 +43,7 @@ type Setup struct {
 // TestAccountWithWalletAndBackend tests an account implementation together with
 // a corresponding wallet and backend implementation.
 // This function should be called by every implementation of the wallet interface.
-func TestAccountWithWalletAndBackend(t *testing.T, s *Setup) {
+func TestAccountWithWalletAndBackend(t *testing.T, s *Setup) { // nolint: revive
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply wallet/test/wallet.go:46]: Added
```pre
//nolint:revive
// func name will be used as test.TestAccountWithWalletAndBackend by other packages,
// and that stutters; consider calling this AccountWithWalletAndBackend (revive).
```

File: watcher/local/watcher.go [hunk 32]
```diff
@@ -100,6 +100,16 @@ type (
...
+// ErrSubChannelsPresent signals that an operation failed because subchannels
+// were present.
+var ErrSubChannelsPresent = stderrors.New("sub-channels present")
+
+// IsErrSubChannelsPresent returns whether the cause of the error was an
+// ErrSubChannelsPresent error.
+func IsErrSubChannelsPresent(err error) bool {
+	return errors.Is(err, ErrSubChannelsPresent)
+}
+
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply watcher/local/watcher.go:112]: Using ad-hoc errors like `stderrors.New("sub-channels present")` is not to the liking of the `goerr113` linter.  
Sadly the value of `stderrors.New` is not a compile-time constant.

File: client/client.go [hunk 33]
```diff
@@ -178,11 +178,11 @@ func (c *Client) Handle(ph ProposalHandler, uh UpdateHandler) {
...
 			go c.handleChannelProposal(ph, env.Sender, msg)
 		case *msgChannelUpdate:
-			go c.handleChannelUpdate(uh, env.Sender, msg)
+			go c.handleChannelUpdate(context.TODO(), uh, env.Sender, msg)
 		case *virtualChannelFundingProposal:
-			go c.handleChannelUpdate(uh, env.Sender, msg)
+			go c.handleChannelUpdate(context.TODO(), uh, env.Sender, msg)
 		case *virtualChannelSettlementProposal:
-			go c.handleChannelUpdate(uh, env.Sender, msg)
+			go c.handleChannelUpdate(context.TODO(), uh, env.Sender, msg)
```
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/client.go:185]: Oh right, this would only make a difference if the client gets closed while there is an on-going request, but i will revert it.

File: wire/net/simple/listener_internal_test.go (hunk 5 above)
Comment (ggwpez) [reply wire/net/simple/listener_internal_test.go:83]: See discussion with the `_ =` above.

**Comment by ggwpez** [comment]
It looks like i messed up some timeout, since a follow up branch for this one fails in the CI…

File: apps/payment/app_internal_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply apps/payment/app_internal_test.go:41]: We talked about it and that was not my conclusion. You may have gotten me wrong. Next time ask if you are unsure and I will try to be more clear.

[the remark this answers is not in the document]
Comment (ggwpez) [reply apps/payment/app_internal_test.go]: Okay anyway, i updated it.  
@matthiasgeihs

File: backend/ethereum/wallet/backend.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply backend/ethereum/wallet/backend.go:37]: Usually I try to go by the following rule: Only expose what is useful to a user of the library. Don't expose things just for the sake of testing. If you need to access internals during testing, you can make it an internal test. Does that make sense?

In this specific case, I have a hard time imagining why a library user would need to access this constant. For testing, you could copy the variable to the test package. What do you think?

File: backend/sim/wallet/address.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply backend/sim/wallet/address.go]: I think you mean they have a shorter string representation. (The byte encoding still holds the full information.)

I am not convinced that cutting off a good chunk of the information for the string representation. In any case, this function could be a one-liner: `return fmt.Sprintf("0x%x...", a.X.Bytes()[:length])`.

File: backend/sim/wallet/address_internal_test.go [hunk 34]
```diff
@@ -45,7 +46,8 @@ func TestAddress_ByteArray(t *testing.T) {
 
 		addr := &Address{
 			X: new(big.Int).SetBytes(dest[:32]),
-			Y: new(big.Int).SetBytes(dest[32:])}
```
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply backend/sim/wallet/address_internal_test.go:48]: Why is the linter not complaining?

File: channel/persistence/keyvalue/restorer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply channel/persistence/keyvalue/restorer.go:95]: Probably that's why: https://golangci-lint.run/usage/configuration/

```
# EXC0001 errcheck: Almost all programs ignore errors on these functions and in most cases it's ok
- Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked
```

File: channel/transaction_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply channel/transaction_test.go]: Instead you can write `tests := [][]bool{}`.

File: client/channelconn.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/channelconn.go]: It seems like enabling the context check linter is more involved. Maybe we postpone it to a separate PR?

File: client/proposal.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/proposal.go:703]: In some cases before we were using context Client.Ctx, and now you are using a different context there. I am not saying that it is necessarily the wrong context you are using. It is just really hard to follow through these changes within such a large PR. I suggest we postpone context linting to a separate PR.

File: client/restore.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/restore.go:27]: I suggest we postpone context linting to a separate PR.

File: client/restore_internal_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/restore_internal_test.go]: :D

File: client/sync.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/sync.go:73]: Seems common practice, hence they adopted the linter. See above more more details.

File: client/sync.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/sync.go]: Would be OK to use nolint here, but also OK to check again, I guess.

File: log/log.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply log/log.go:56]: I was assuming something like that. Could you extend the function docu respectively?

File: wallet/sig.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply wallet/sig.go]: I think this is another instance where using `//nolint` would be justified. Anyways, also OK like this.

File: wire/net/simple/listener_internal_test.go (hunk 5 above)
Comment (matthiasgeihs) [reply wire/net/simple/listener_internal_test.go]: See answer above.

File: client/client.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply client/client.go]: I suggest we postpone context linting to a separate PR.

File: wire/net/simple/listener_internal_test.go (hunk 5 above)
Comment (ggwpez) [reply wire/net/simple/listener_internal_test.go]: Already updated it in the last iteration.

File: client/client.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/client.go]: Yes i also did not like the ctx changes, will remove the liner for now. Otherwise it wants to have `nolint` at every call in the tree where a ctx could be passed, that is annoying as well.  
If we instead remove the ctx from the `relay.Cache` it would be fixed.

File: backend/sim/wallet/address.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/sim/wallet/address.go]: I put it in https://github.com/hyperledger-labs/go-perun/issues/261

File: channel/persistence/keyvalue/restorer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply channel/persistence/keyvalue/restorer.go:95]: Ok interesting. Then i close the discussion.

File: client/channelconn.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/channelconn.go]: Yes i moved it out.

File: channel/transaction_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply channel/transaction_test.go]: I thought this would cause another linter to fail, but it works.

File: client/proposal.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply client/proposal.go]: Factored the context changes out.

File: log/log.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply log/log.go:56]: I dont understand why the linters dont fail here as well, since it are undocumented interface functions :nauseated_face:  
Created issue https://github.com/hyperledger-labs/go-perun/issues/262

File: backend/ethereum/wallet/backend.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply backend/ethereum/wallet/backend.go:37]: I copied the variable and unexported it.

**Comment by ggwpez** [comment]
> It looks like i messed up some timeout, since a follow up branch for this one fails in the CI…

Could not reproduce now after ~8 runs, maybe was false alarm with slow runners.  
The run in question was [this](https://github.com/perun-network/go-perun/actions/runs/1472566878/attempts/1).

**Comment by matthiasgeihs** [comment]
> > It looks like i messed up some timeout, since a follow up branch for this one fails in the CI…
> 
> Could not reproduce now after ~8 runs, maybe was false alarm with slow runners. The run in question was [this](https://github.com/perun-network/go-perun/actions/runs/1472566878/attempts/1).

Might have been related to the context changes?

File: wallet/test/address.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply wallet/test/address.go]: I meant rather like `//nolint // explanation`. There is also a similar example here https://golangci-lint.run/usage/false-positives/.

Furthermore, `//nolint: revive` should apply only to the respective line, not to the whole function. You can write the following:

```go
func TestAccountWithWalletAndBackend(t *testing.T, s *Setup) { //nolint: revive // `test.Test...` stutters, but we accept that here.
```

File: wallet/test/wallet.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (matthiasgeihs) [reply wallet/test/wallet.go]: See https://github.com/hyperledger-labs/go-perun/pull/260#discussion_r752285107.

File: backend/ethereum/channel/funder.go [hunk 35]
```diff
@@ -211,7 +213,7 @@ func (f *Funder) fundAssets(ctx context.Context, channelID channel.ID, req chann
 // funding request. It is idempotent.
 func (f *Funder) sendFundingTx(ctx context.Context, request channel.FundingReq, contract assetHolder, fundingID [32]byte) (txs []*types.Transaction, fatal error) {
 	bal := request.Agreement[contract.assetIndex][request.Idx]
-	// nolint: gocritic
+	//nolint:gocritic
```
Comment (matthiasgeihs) [inline backend/ethereum/channel/funder.go:216]: This `//nolint` can be avoided by improving the structure:

```go
// sendFundingTx sends and returns the TXs that are needed to fulfill the
// funding request. It is idempotent.
func (f *Funder) sendFundingTx(ctx context.Context, request channel.FundingReq, contract assetHolder, fundingID [32]byte) (txs []*types.Transaction, fatal error) {
	bal := request.Agreement[contract.assetIndex][request.Idx]

	if bal == nil || bal.Sign() <= 0 {
		f.log.WithFields(log.Fields{"channel": request.Params.ID(), "idx": request.Idx}).Debug("Skipped zero funding.")
		return nil, nil
	}

	alreadyFunded, err := f.checkFunded(ctx, bal, contract, fundingID)
	if err != nil {
		return nil, errors.WithMessage(err, "checking funded")
	}

	if alreadyFunded {
		f.log.WithFields(log.Fields{"channel": request.Params.ID(), "idx": request.Idx}).Debug("Skipped second funding.")
		return nil, nil
	}

	return f.deposit(ctx, bal, wallet.Address(*contract.Address), fundingID)
}
```

**Review by matthiasgeihs — COMMENTED** [review]
I couldn't check all the files again. I am assuming it is OK. Just a few open comments left.

File: wallet/test/address.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply wallet/test/address.go]: Weird, i tried that before as well but it did not work.

File: wallet/test/wallet.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (ggwpez) [reply wallet/test/wallet.go]: updated

File: backend/ethereum/channel/funder.go (hunk 35 above)
Comment (ggwpez) [reply backend/ethereum/channel/funder.go]: put it in https://github.com/hyperledger-labs/go-perun/pull/260/commits/cdc57f576b60eb39e266ad6c3200d500c519d2f3

**Review by matthiasgeihs — APPROVED** [review]
